pss/crypto/crypto.go

```diff
@@ -0,0 +1,656 @@
+	WrapMessage(msgData []byte, params *WrapParams) (data []byte, err error)
+	UnWrapSymmetric(encryptedData, key []byte) (ReceivedMessage, error)
+	UnWrapAsymmetric(encryptedData []byte, key *ecdsa.PrivateKey) (ReceivedMessage, error)
+}
+
+// CryptoBackend contains key manipulation methods
+type CryptoBackend interface {
+
+	// Key store functions
+	GetSymKey(id string) ([]byte, error)
+	GenerateSymKey() (string, error)
+	AddSymKeyDirect(bytes []byte) (string, error)
+
+	// Key conversion
+	FromECDSAPub(pub *ecdsa.PublicKey) []byte
+
+	// Key serialization
+	UnmarshalPubkey(pub []byte) (*ecdsa.PublicKey, error)
+	CompressPubkey(pubkey *ecdsa.PublicKey) []byte
+}
+
+// CryptoUtils contains utility methods for tests
+type CryptoUtils interface {
+	GenerateKey() (*ecdsa.PrivateKey, error)
+	NewKeyPair(ctx context.Context) (string, error)
+	GetPrivateKey(id string) (*ecdsa.PrivateKey, error)
+}
+
+var (
+	errInvalidPubkey               = errors.New("invalid public key provided for asymmetric encryption")
+	errInvalidSymkey               = errors.New("invalid key provided for symmetric encryption")
+	errMissingSaltOrInvalidPayload = errors.New("missing salt or invalid payload in symmetric message")
+	errSecureRandomData            = errors.New("failed to generate secure random data")
+	errNeitherSymNorAsymKeysProv   = errors.New("unable to encrypt the message: neither symmetric nor asymmetric key provided")
+)
+
+type defaultCryptoBackend struct {
+	privateKeys map[string]*ecdsa.PrivateKey // Private key storage
+	symKeys     map[string][]byte            // Symmetric key storage
+	keyMu       sync.RWMutex                 // Mutex associated with key storages
+}
+
+// receivedMessage represents a data packet to be received
+// and successfully decrypted.
+type receivedMessage struct {
+	Payload   []byte // Parsed and validated message content
+	Raw       []byte // Unparsed but decrypted message content
+	Signature []byte // Signature from the sender
+	Salt      []byte // Salt used for validation
+	Padding   []byte // Padding applied
+
+	Src *ecdsa.PublicKey // Source public key used for signing the message
+	Dst *ecdsa.PublicKey // Destination public key used for encrypting the msessage asymmetrically
+
+	crypto *defaultCryptoBackend
+}
+
+func (msg receivedMessage) GetSrc() *ecdsa.PublicKey {
+	return msg.Src
+}
+
+// ValidateAndParse checks that the format and the signnature are correct. It also set Payload as the parsed message
+func (msg *receivedMessage) ValidateAndParse() bool {
+	end := len(msg.Raw)
+	if end < 1 {
+		return false
+	}
+	if isMessageSigned(msg.Raw[0]) {
+		end -= signatureLength
+		if end <= 1 {
+			return false
+		}
+		msg.Signature = msg.Raw[end : end+signatureLength]
+		msg.Src = msg.sigToPubKey()
+		if msg.Src == nil {
+			return false
+		}
+	}
+
+	beg := 1
+	payloadSize := 0
+	sizeOfPayloadSizeField := int(msg.Raw[0] & SizeMask) // number of bytes indicating the size of payload
+	if sizeOfPayloadSizeField != 0 {
```

Comment:
Please consider whether payload of 0 size should be allowed, or if it should generate an error.

Reply:
Note that here what we are checking is the size of `PayloadSize` field. That field should be at least size 1, even when the `payloadSize` is 0 (one byte to represent 0).